core/service/src/lib.rs

```diff
@@ -222,6 +222,23 @@ impl<Components: components::Components> Service<Components> {
 			task_executor.spawn(events);
 		}
 
+		{
+			// finality notifications
+			let network = Arc::downgrade(&network);
+
+			let events = client.finality_notification_stream()
+				.for_each(move |notification| {
```

Comment:
since we are currently always calling `finalize_block(..., true)` on `client`, that means that we are going to produce a ton of finality notifications. Perhaps it should be a custom future which does

```rust
struct MostRecentNotification(Fuse<notifications>);

fn poll() -> _ {
    let mut last = None;
    let last = loop {
        match self.0.poll()? {
             Async::Ready(Some(item)) => { last = Some(item) }
             // safe because fused.
             Async::Ready(None) => if last.is_none() { return Ok(Async::Ready(None)) } else { break last }
             Async::NotReady => break last,
         }
    };

    Ok(Async::Ready(Some(last))
}
```